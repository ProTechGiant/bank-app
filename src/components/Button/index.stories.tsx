import { expect } from "@storybook/jest";
import { ComponentMeta, ComponentStory } from "@storybook/react";
import { userEvent, within } from "@storybook/testing-library";

import Button from "./index";

export default {
  title: "components/Button",
  component: Button,
  args: {
    children: "Hello World",
  },
  argTypes: {
    disabled: {
      control: "boolean",
    },
    iconLeft: {
      table: {
        disable: true,
      },
    },
    iconRight: {
      table: {
        disable: true,
      },
    },
    onPress: {
      action: "onPress",
      table: {
        disable: true,
      },
    },
  },
} as ComponentMeta<typeof Button>;

export const TextButton: ComponentStory<typeof Button> = args => <Button {...args} />;
TextButton.args = { variant: "primary" };
TextButton.play = async ({ args, canvasElement }) => {
  const canvas = within(canvasElement);

  await userEvent.click(canvas.getByText(args.children as string));
  await expect(args.onPress).toHaveBeenCalled();
};

export const LoadingButton: ComponentStory<typeof Button> = args => <Button {...args}>Loading button</Button>;
LoadingButton.args = { type: "loader" };
