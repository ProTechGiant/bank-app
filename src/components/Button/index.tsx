import { ActivityIndicator, Pressable, PressableProps, StyleSheet, View, ViewStyle } from "react-native";

import Typography from "@/components/Typography";
import { useThemeStyles } from "@/theme";

interface ButtonProps extends Omit<PressableProps, "children" | "disabled" | "style"> {
  block?: boolean;
  children?: string | React.ReactNode;
  color?: "base" | "alt";
  disabled?: boolean;
  iconLeft?: React.ReactNode;
  iconRight?: React.ReactNode;
  type?: "has icons" | "no icons" | "loader";
  variant?: "primary" | "secondary" | "tertiary";
}

export default function Button({
  block = false,
  children,
  color = "base",
  iconLeft,
  disabled,
  iconRight,
  type = "no icons",
  variant = "primary",
  ...restProps
}: ButtonProps) {
  const containerStyles = useThemeStyles<ViewStyle>(
    theme => ({
      backgroundColor:
        variant === "primary"
          ? color === "base"
            ? theme.palette.primaryBase
            : theme.palette.complimentBase
          : undefined,
      borderRadius: theme.radii.extraSmall,
      borderWidth: variant === "secondary" ? 1 : 0,
      borderColor: color === "base" ? theme.palette.primaryBase : theme.palette.complimentBase,
      paddingHorizontal: type === "has icons" ? theme.spacing["8p"] : theme.spacing["16p"],
      paddingVertical: theme.spacing["16p"],
    }),
    [type, variant]
  );

  return (
    <Pressable {...restProps} disabled={disabled} style={block && styles.containerBlock}>
      <View style={[styles.container, containerStyles, { opacity: disabled ? 0.5 : 1 }]}>
        {undefined !== iconLeft && <View style={styles.icon}>{iconLeft}</View>}
        {type === "loader" ? (
          <ActivityIndicator color="white" size="small" />
        ) : (
          <Typography.Text
            color={variant === "primary" ? "neutralBase-30" : color === "base" ? "primaryBase" : "complimentBase"}
            size="body"
            weight="regular">
            {children}
          </Typography.Text>
        )}
        {undefined !== iconRight && <View style={styles.icon}>{iconRight}</View>}
      </View>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    display: "flex",
    flexDirection: "row",
    justifyContent: "center",
  },
  containerBlock: {
    alignSelf: "stretch",
  },
  icon: {
    marginHorizontal: 8,
  },
});
