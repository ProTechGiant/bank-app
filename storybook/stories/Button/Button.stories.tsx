import { storiesOf } from "@storybook/react-native";
import { Text } from "react-native";

import CenterView from "../CenterView";
import Button from ".";

storiesOf("Button", module)
  .addDecorator(getStory => <CenterView>{getStory()}</CenterView>)
  .add("with text", () => (
    <Button>
      <Text>Hello there!</Text>
    </Button>
  ))
  .add("with some emoji", () => (
    <Button>
      <Text>😀 😎 👍 💯</Text>
    </Button>
  ));
