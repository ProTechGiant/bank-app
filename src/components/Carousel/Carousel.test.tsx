import { fireEvent, render, screen } from "@testing-library/react-native";
import { Text, View } from "react-native";
import { create } from "react-test-renderer";

import Button from "../Button";
import { Carousel } from "./index";

const data = [
  {
    text: "Slide 1",
  },
  {
    text: "Slide 2",
  },
  {
    text: "Slide 3",
  },
];

const TestCarousel = () => {
  return (
    <View>
      <Carousel data={data} Slide={CarouselSlide} pagination="overlay" />
      <Text testID="printed-text">test</Text>
    </View>
  );
};

const CarouselSlide = ({ data }: { data: { text: string } }) => (
  <View testID="slide">
    <Text>{data.text}</Text>
  </View>
);

describe("components/Carousel", () => {
  it("renders a Carousel", () => {
    const result = create(<Carousel data={data} Slide={CarouselSlide} pagination="overlay" />);

    expect(result).toMatchSnapshot();
  });

  it("should render Slides", async () => {
    render(<TestCarousel />);
    const slides = await screen.getAllByTestId("slide");

    expect(slides.length).toBe(3);
  });
});
