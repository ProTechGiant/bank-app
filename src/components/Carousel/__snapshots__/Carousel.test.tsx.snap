// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Carousel renders a Carousel 1`] = `
<View
  style={
    Object {
      "paddingTop": 8,
      "position": "relative",
      "width": "100%",
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "text": "Slide 1",
        },
        Object {
          "text": "Slide 2",
        },
        Object {
          "text": "Slide 3",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    pagingEnabled={true}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={16}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    snapToAlignment="center"
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "width": undefined,
            }
          }
        >
          <View
            testID="slide"
          >
            <Text>
              Slide 1
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "width": undefined,
            }
          }
        >
          <View
            testID="slide"
          >
            <Text>
              Slide 2
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "width": undefined,
            }
          }
        >
          <View
            testID="slide"
          >
            <Text>
              Slide 3
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "bottom": 8,
          "flexDirection": "row",
          "position": "absolute",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignSelf": "center",
              "borderRadius": 4,
              "marginBottom": 8,
              "marginHorizontal": 2,
            },
            Object {
              "backgroundColor": "#FF7512",
              "height": 8,
              "width": 8,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignSelf": "center",
              "borderRadius": 4,
              "marginBottom": 8,
              "marginHorizontal": 2,
            },
            Object {
              "backgroundColor": "#CFCFDD",
              "height": 6,
              "width": 6,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignSelf": "center",
              "borderRadius": 4,
              "marginBottom": 8,
              "marginHorizontal": 2,
            },
            Object {
              "backgroundColor": "#CFCFDD",
              "height": 6,
              "width": 6,
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
